pkg/deploy/factory.go

```diff
@@ -38,32 +41,42 @@ func NewFactory(coreClient kubernetes.Interface,
 		NewKubeconfigSecrets(coreClient), NewServiceAccounts(coreClient, log), cmdRunner}
 }
 
+// NewKapp configures and returns a deployer of type Kapp
 func (f Factory) NewKapp(opts v1alpha1.AppDeployKapp, saName string,
 	clusterOpts *v1alpha1.AppCluster, genericOpts GenericOpts, cancelCh chan struct{}) (*Kapp, error) {
 
+	processedGenericOpts, err := processOpts(saName, clusterOpts, genericOpts, f.serviceAccounts, f.kubeconfigSecrets)
+	if err != nil {
+		return nil, err
+	}
+	const suffix string = ".app"
+	return NewKapp(suffix, opts, *processedGenericOpts,
+		f.globalKappDeployRawOpts(), cancelCh, f.cmdRunner), nil
+}
+
+// processOpts takes generic opts and a ServiceAccount Name, and returns a populated kubeconfig that can connect to a cluster.
+// if the saName is empty then you'll connect to a cluster via the clusterOpts inside the genericOpts, otherwise you'll use the specified SA.
+func processOpts(saName string, clusterOpts *v1alpha1.AppCluster, genericOpts GenericOpts, serviceAccounts *ServiceAccounts, kubeconfigSecrets *KubeconfigSecrets) (*ProcessedGenericOpts, error) {
```

Comment:
since we have placed GetClusterVersion onto Factory, we should be able to reduce processOpts scope as well but placing it onto Factory. by doing that it will also not have to accept serviceAccounts and kubeconfigSecrets and just read off f.